src/NuGet.Core/NuGet.PackageManagement/Resolution/PrunePackageTree.cs

```diff
@@ -63,30 +63,32 @@ private static void CollectAllowedFromDependenciesOfPackagesToInstall(IEnumerabl
         {
             var prereleasePackageToInstall = new HashSet<PackageIdentity>(packagesToInstall.Where(p => p.HasVersion && p.Version.IsPrerelease), PackageIdentityComparer.Default);
 
-            var maxDepth = packages.Select(p => p.Id).Distinct(StringComparer.OrdinalIgnoreCase).Count();
+            var visitedNodes = new HashSet<SourcePackageDependencyInfo>();
 
             foreach (var packageToInstall in packages.Where(p => prereleasePackageToInstall.Contains(p)))
             {
-                int depth = 0;
-                WalkDependencies(packages, packageToInstall, allowed, depth, maxDepth);
+                WalkDependencies(packages, packageToInstall, allowed, visitedNodes);
             }
         }
 
-        private static void WalkDependencies(IEnumerable<SourcePackageDependencyInfo> packages, SourcePackageDependencyInfo packageToInstall, HashSet<string> allowed, int depth, int maxDepth)
+        private static void WalkDependencies(IEnumerable<SourcePackageDependencyInfo> packages, SourcePackageDependencyInfo packageToInstall, HashSet<string> allowed, HashSet<SourcePackageDependencyInfo> visitedNodes)
         {
-            if (depth > maxDepth)
-            {
-                return;
-            }
+            var queue = new Queue<SourcePackageDependencyInfo>();
+            queue.Enqueue(packageToInstall);
 
-            if (packageToInstall.Dependencies != null)
+            while (queue.Count > 0)
             {
-                foreach (var dependency in packageToInstall.Dependencies)
+                var package = queue.Dequeue();
+                foreach (var dependency in package.Dependencies)
                 {
                     allowed.Add(dependency.Id);
                     foreach (SourcePackageDependencyInfo dependentPackage in packages.Where(p => p.Id.Equals(dependency.Id, StringComparison.OrdinalIgnoreCase)))
                     {
-                        WalkDependencies(packages, dependentPackage, allowed, depth + 1, maxDepth);
+                        if (!visitedNodes.Any(p => p.Equals(dependentPackage)))
```

Comment:
Could this be visitedNodes.Contains(dependentPackage)? It should be much faster to check a hashset that way